Add activate/drain/pause node command

vdemeester opened the pull request:
Expose simple commands for node update to be more consistent with the current ones (`demote`/`promote`, `accept`/`reject`) 🐮.

Closes #24086

/cc @aluzzardi @aaronlehmann @cpuguy83 @dnephin 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

tonistiigi commented on docs/reference/commandline/node_activate.md:
```diff
@@ -0,0 +1,27 @@
++++
+<![end-metadata]-->
+
+# node accept
```
`s/accept/activate/` + other files as well

vdemeester replied:
ah damn 😅

aaronlehmann commented on api/client/node/pause.go:
```diff
@@ -0,0 +1,31 @@
+		node.Spec.Availability = swarm.NodeAvailabilityPause
+	}
+	success := func(nodeID string) {
+		fmt.Fprintf(dockerCli.Out(), "Node %s availability set up to pause.\n", nodeID)
```
s/set up to/set to/

aaronlehmann commented on docs/reference/commandline/node_activate.md:
```diff
@@ -0,0 +1,27 @@
+
+    Activate a node in the swarm
+
+Activate a node into the swarm. This command targets a docker engine that is a manager in the swarm cluster.
```
Let's explain what it means to activate a node. It means that we're taking a node that was paused or drained, and changing it so it's no longer paused and drained, so it can start accepting task assignments.

> This command targets a docker engine that is a manager in the swarm cluster.

This may be a little unclear - it might sound as if you are activating a manager. We could say "This command must be run on a manager node, but may activate any node in the swarm."

aaronlehmann replied:
Also, s/into/in/.

aaronlehmann commented on docs/reference/commandline/node_activate.md:
````diff
@@ -0,0 +1,27 @@
+
+
+```bash
+$ docker node activate <node name>
````
Or node ID, right? Is there a concise way to show that either works? Maybe another line with

```bash
docker node activate <node ID>
```

tonistiigi replied:
There is no node name(you can't inspect by hostname). It's ID or prefix.

aaronlehmann commented on docs/reference/commandline/node_drain.md:
```diff
@@ -0,0 +1,27 @@
+
+    Drain a node in the swarm
+
+Drain a node into the swarm. This command targets a docker engine that is a manager in the swarm cluster.
```
Explain what draining means: all tasks will be removed from this node, and no tasks will be assigned to it until it is reactivated with "docker node activate".

Same comment as above about "This command targets a docker engine...".

aaronlehmann replied:
Also, s/into/in/.

aaronlehmann commented on docs/reference/commandline/node_pause.md:
```diff
@@ -0,0 +1,27 @@
+
+    Pause a node in the swarm
+
+Pause a node into the swarm. This command targets a docker engine that is a manager in the swarm cluster.
```
Explain what pausing means: no new tasks will be assigned to this node until it is reactivated with "docker node activate".

Same comment as above about "This command targets a docker engine...".

s/into/in/

aaronlehmann commented on api/client/node/drain.go:
```diff
@@ -0,0 +1,31 @@
+		node.Spec.Availability = swarm.NodeAvailabilityDrain
+	}
+	success := func(nodeID string) {
+		fmt.Fprintf(dockerCli.Out(), "Node %s availability set up to drain.\n", nodeID)
```
s/set up to/set to/

aaronlehmann commented on api/client/node/activate.go:
```diff
@@ -0,0 +1,31 @@
+		node.Spec.Availability = swarm.NodeAvailabilityActive
+	}
+	success := func(nodeID string) {
+		fmt.Fprintf(dockerCli.Out(), "Node %s availability set up to active.\n", nodeID)
```
s/set up to/set to/

aaronlehmann commented on api/client/node/activate.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func runActivate(dockerCli *client.DockerCli, nodes []string) error {
+	accept := func(node *swarm.Node) {
```
Rename the closure to `activate`.

aaronlehmann commented on api/client/node/drain.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func runDrain(dockerCli *client.DockerCli, nodes []string) error {
+	accept := func(node *swarm.Node) {
```
Rename the closure to `drain`.

aaronlehmann commented on api/client/node/pause.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func runPause(dockerCli *client.DockerCli, nodes []string) error {
+	accept := func(node *swarm.Node) {
```
Rename the closure to `pause`.

vdemeester commented:
@aaronlehmann @tonistiigi upadted (alongside with other `node_*.md` docs)

aaronlehmann commented on docs/reference/commandline/node_accept.md:
```diff
@@ -14,7 +14,11 @@ parent = "smn_cli"
+able to accept tasks.
+
+This command must be run on a manager node, but may accept any node in
+the swarm.
```
s/may accept any node in the swarm/may accept any pending node/

aaronlehmann commented on docs/reference/commandline/node_activate.md:
```diff
@@ -0,0 +1,30 @@
+
+    Activate a node in the swarm
+
+This command mark a node previously drained or paused as active again
```
s/mark a node previously drained or paused as active again/marks a previously drained or paused node as active/

aaronlehmann commented on docs/reference/commandline/node_drain.md:
```diff
@@ -0,0 +1,31 @@
+
+    Drain a node in the swarm
+
+This command mark a node as drained so that all tasks will be removed
```
s/mark/marks/

aaronlehmann commented on docs/reference/commandline/node_pause.md:
```diff
@@ -0,0 +1,30 @@
+
+    Pause a node in the swarm
+
+This command mark a node as paused so that it doesn't accept any new
```
s/mark/marks/

aaronlehmann commented on docs/reference/commandline/node_pause.md:
```diff
@@ -0,0 +1,30 @@
+    Pause a node in the swarm
+
+This command mark a node as paused so that it doesn't accept any new
+tasks until it's been reactivade (using `docker node activate`).
```
s/reactivade/reactivated/

aaronlehmann commented on docs/reference/commandline/node_update.md:
```diff
@@ -21,7 +21,13 @@ parent = "smn_cli"
       --role string           Role of the node (worker/manager)
 
+This command is a *low-level* command to update membership, role and
+availability of a node. Each of this *attribute* can be updated using
```
"Each of these *attributes* can be updated using a simpler command"

aaronlehmann commented on docs/reference/commandline/node_promote.md:
```diff
@@ -14,8 +14,10 @@ parent = "smn_cli"
     Promote a node as manager in the swarm
 
-Promotes a node that is pending a promotion to manager. This command targets a docker engine that is a manager in the swarm cluster.
+Promotes a node that is pending a promotion to manager.
```
Not related to the change in this PR, but I noticed this is wrong. There is no such thing as pending a promotion to manager. It should say:

 "Promotes a worker node to a manager. Becoming a manager means the node will be able to accept administrative commands for the swarm, and become one of the candidates when the swarm elects a leader. The node will continue to be assigned tasks unless it is paused with `docker node pause` or drained with `docker node drain`. Only specific, stable nodes should be promoted to managers, because a majority of leaders must be responsive for the swarm to be available. Generally, a swarm should be set up with either 1, 3, or 5 managers."

vdemeester replied:
I well update this too (as I already modified the file :wink:)

vdemeester commented:
Updated again :wink: 
Thanks @aaronlehmann :+1:

aaronlehmann commented:
LGTM

Thanks for making all the changes.

dnephin commented:
Changes LGTM, but I'm still concerned this is the wrong trade-off to make here if we want any hope of being able to expand the definition of  a "node" object to being larger than just a "swarm node".  Adding a bunch of swarm node "state" specific commands is going to lead to a confusing and bloated UI for node.

We don't have these kind of trivial commands anywhere else in the CLI, we generally have a command with flags (`docker update`, `docker service update`).

thaJeztah commented:
I'm adding this to the UX document for discussion

cpuguy83 commented:
@dnephin I'm not sure why we would have a concept of a `node` outside of swarm.

cpuguy83 commented:
@vdemeester Needs a rebase, but otherwise LGTM.

vdemeester commented:
@cpuguy83 rebased 👼

thaJeztah commented on docs/reference/commandline/node_update.md:
```diff
@@ -23,6 +23,14 @@ Options:
+This command is a *low-level* command to update membership, role and
+availability of a node. Each of these attributes can be updated using
+a simpler command such as `accept` for membership, `promote` or
+`demote` for role and `activate`, `drain` or `pause` for
```
these should probably be links to those commands

vdemeester replied:
Adding the links into `Related information` :wink:

thaJeztah replied:
`accept` is no longer an option, so needs to be removed if we go forward

vdemeester replied:
ah right :angel:

thaJeztah commented on docs/reference/commandline/node_update.md:
```diff
@@ -23,6 +23,14 @@ Options:
+This command must be run on a manager node, but may update any node in
+the swarm.
 
 ## Related information
```
oh, and we need to update all the "related" sections to include the new commands

thaJeztah commented:
docs LGTM

cpuguy83 commented:
Moving back to design review, b/c unfortunately this is not 100% agreed upon.

icecrime commented:
I won't vote for/against here, but I'll vote for _one_ thing: not rushing this decision for 1.12.0 :wink: I'll remove from the milestone until we can get consensus, sorry!

aaronlehmann commented:
I vote that we include this in 1.12. What needs to happen to reach a consensus?

vdemeester commented:
I tend to agree with @icecrime, let's not *rush* things. It's better to not have them in 1.12 and have them or not in 1.13 than adding them and backing up afterwards :angel:.

@thaJeztah @icecrime  should we brought this one in a maintainer session (and adds a `needs-attention` label) ?

LK4D4 commented:
@vdemeester @aaronlehmann @thaJeztah it's time! I'm okay with design.

aaronlehmann commented:
@vdemeester: Please rebase :)

aluzzardi commented:
Actually, quite the opposite - I would vote for removing `promote` and `demote` and not add porcelain commands. `--availability=[drain|pause|active]` is quite easy already.

WDYT?

aaronlehmann commented:
I'm not in favor of removing commands right after the release.

vdemeester commented:
@aluzzardi the least would be to deprecate them (and thus keeping them for at least 3 releases). I do agree with @aaronlehmann on not removing them (`promote`/`demote`) — and not rush on deprecate them either :angel: (but this can be subject of debate).

vdemeester commented:
Rebased. If we decide not to move forward, I'll re-open a PR for the documentation changes that are on `node_promote.md` and `node_update.md` 👼

vdemeester committed:
Add activate/drain/pause node command

Expose simple commands for node update to be more consistent with the
current ones (demote/promote, accept/reject).

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
let's close this PR for now, and discuss on the issue which states we consider frequently used, and want a porcelain version for.

We can reopen after that https://github.com/docker/docker/issues/24086
